New typings broke custom elements for me

The new typings introduced in #1116 broke custom elements for me. Is there a workaround, am I using `JSX.IntrinsicElements` wrong or do we need a fix for preact’s typings?

Test case (attach this to `test/ts/VNode-test.s`):

```typescript
class TestCE extends HTMLElement {}

declare global {
  namespace JSX {
    interface IntrinsicElements {
      ["test-ce"]: Partial<TestCE>;
    }
  }
}

const MyTestCustomElement = <test-ce><div>hai</div></test-ce>;
```

Error:

```
test/ts/VNode-test.tsx:78:30 - error TS2322: Type '{ children: Element; }' is not assignable to type 'Partial<TestCE>'.
  Types of property 'children' are incompatible.
    Type 'Element' is not assignable to type 'HTMLCollection | undefined'.
      Type 'Element' is not assignable to type 'HTMLCollection'.
        Property 'namedItem' is missing in type 'Element'.

const MyTestCustomElement = <test-ce><div>hai</div></test-ce>;
```
cc @marvinhagemeister

Sadly, TypeScript seems too smart there, too 😅 

```ts
const PinchZoom = "pinch-zoom"

export default class Inspector extends Component {
  render() {
    return (
      <PinchZoom/>
    );
  }
}
```

Errors with

```
src/components/inspector/index.tsx:15:7 - error TS2339: Property 'pinch-zoom' does not exist on type 'JSX.IntrinsicElements'.

15       <PinchZoom />
         ~~~~~~~~~~~~~

src/components/inspector/index.tsx:15:8 - error TS2604: JSX element type 'PinchZoom' does not have any construct or call signatures.

15       <PinchZoom />
          ~~~~~~~~~
```

--- a/package.json
+++ b/package.json
@@ -230,7 +230,7 @@
     "prop-types": "^15.7.2",
     "sinon": "^6.1.3",
     "sinon-chai": "^3.0.0",
-    "typescript": "^3.0.1",
+    "typescript": "3.5.3",
     "webpack": "^4.3.0"
   }
 }

--- a/src/index.d.ts
+++ b/src/index.d.ts
@@ -15,7 +15,7 @@ declare namespace preact {
 		props: P & { children: ComponentChildren };
 		key: Key;
 		/**
-		 * ref is not guaranteed by React.ReactElement, for compatiblity reasons
+		 * ref is not guaranteed by React.ReactElement, for compatibility reasons
 		 * with popular react libs we define it as optional too
 		 */
 		ref?: Ref<any> | null;

--- a/test/ts/Component-test.tsx
+++ b/test/ts/Component-test.tsx
@@ -1,11 +1,6 @@
 import 'mocha';
 import { expect } from 'chai';
-import {
-	createElement,
-	Component,
-	RenderableProps,
-	Fragment
-} from '../../src/';
+import { createElement, Component, RenderableProps, Fragment } from '../../';
 
 // Test `this` binding on event handlers
 function onHandler(this: HTMLInputElement, event: any) {

--- a/test/ts/VNode-test.tsx
+++ b/test/ts/VNode-test.tsx
@@ -10,7 +10,7 @@ import {
 	VNode,
 	ComponentChildren,
 	cloneElement
-} from '../../src';
+} from '../../';
 
 function getDisplayType(vnode: VNode | string | number) {
 	if (typeof vnode === 'string' || typeof vnode == 'number') {

--- a/test/ts/custom-elements.tsx
+++ b/test/ts/custom-elements.tsx
@@ -5,6 +5,7 @@ declare module '../../' {
 		interface IntrinsicElements {
 			// Custom element can use JSX EventHandler definitions
 			'clickable-ce': {
+				optionalAttr?: string;
 				onClick?: MouseEventHandler<HTMLElement>;
 			};
 
@@ -45,17 +46,28 @@ const { Provider, Consumer } = createContext({ contextValue: '' });
 // Sample component that uses custom elements
 
 class SimpleComponent extends Component {
+	componentProp = 'componentProp';
 	render() {
 		// Render inside div to ensure standard JSX elements still work
 		return (
 			<Provider value={{ contextValue: 'value' }}>
 				<div>
 					<clickable-ce
-						onClick={e => console.log('clicked ', e.target)}
+						onClick={e => {
+							// `this` should be instance of SimpleComponent since this is an
+							// arrow function
+							console.log(this.componentProp);
+
+							// Validate `currentTarget` is HTMLElement
+							console.log('clicked ', e.currentTarget.style.display);
+						}}
 					></clickable-ce>
-					<color-picker space="rgb"></color-picker>
+					<color-picker space="rgb" dir="rtl"></color-picker>
 					<custom-whatever
+						dir="auto" // Inherited prop from HTMLAttributes
+						someattribute="string"
 						onsomeevent={function(e) {
+							// Validate `this` and `e` are the right type
 							console.log('clicked', this.instanceProp, e.eventProp);
 						}}
 					></custom-whatever>

--- a/test/ts/custom-elements.tsx
+++ b/test/ts/custom-elements.tsx
@@ -1,0 +1,73 @@
+import { createElement, Component, createContext } from '../../';
+
+declare module '../../' {
+	namespace createElement.JSX {
+		interface IntrinsicElements {
+			// Custom element can use JSX EventHandler definitions
+			'clickable-ce': {
+				onClick?: MouseEventHandler<HTMLElement>;
+			};
+
+			// Custom Element that extends HTML attributes
+			'color-picker': HTMLAttributes & {
+				// Required attribute
+				space: 'rgb' | 'hsl' | 'hsv';
+				// Optional attribute
+				alpha?: boolean;
+			};
+
+			// Custom Element with custom interface definition
+			'custom-whatever': WhateveElAttributes;
+		}
+	}
+}
+
+// Whatever Element definition
+
+interface WhateverElement {
+	instanceProp: string;
+}
+
+interface WhateverElementEvent {
+	eventProp: number;
+}
+
+// preact.JSX.HTMLAttributes also appears to work here but for consistency,
+// let's use createElement.JSX
+interface WhateveElAttributes extends createElement.JSX.HTMLAttributes {
+	someattribute?: string;
+	onsomeevent?: (this: WhateverElement, ev: WhateverElementEvent) => void;
+}
+
+// Ensure context still works
+const { Provider, Consumer } = createContext({ contextValue: '' });
+
+// Sample component that uses custom elements
+
+class SimpleComponent extends Component {
+	render() {
+		// Render inside div to ensure standard JSX elements still work
+		return (
+			<Provider value={{ contextValue: 'value' }}>
+				<div>
+					<clickable-ce
+						onClick={e => console.log('clicked ', e.target)}
+					></clickable-ce>
+					<color-picker space="rgb"></color-picker>
+					<custom-whatever
+						onsomeevent={function(e) {
+							console.log('clicked', this.instanceProp, e.eventProp);
+						}}
+					></custom-whatever>
+
+					{/* Ensure context still works */}
+					<Consumer>
+						{({ contextValue }) => contextValue.toLowerCase()}
+					</Consumer>
+				</div>
+			</Provider>
+		);
+	}
+}
+
+const component = <SimpleComponent />;

--- a/test/ts/hoc-test.tsx
+++ b/test/ts/hoc-test.tsx
@@ -4,7 +4,7 @@ import {
 	ComponentFactory,
 	ComponentConstructor,
 	Component
-} from '../../src';
+} from '../../';
 import { SimpleComponent, SimpleComponentProps } from './Component-test';
 
 export interface highlightedProps {

--- a/test/ts/jsx-namespacce-test.tsx
+++ b/test/ts/jsx-namespacce-test.tsx
@@ -1,4 +1,4 @@
-import { createElement, Component } from '../../src';
+import { createElement, Component } from '../../';
 
 // declare global JSX types that should not be mixed with preact's internal types
 declare global {

--- a/test/ts/preact.tsx
+++ b/test/ts/preact.tsx
@@ -5,7 +5,7 @@ import {
 	FunctionalComponent,
 	AnyComponent,
 	h
-} from '../../src';
+} from '../../';
 
 interface DummyProps {
 	initialInput: string;
@@ -207,7 +207,8 @@ class DefaultPropsWithUnion extends Component<
 		| {
 				type: 'number';
 				num: number;
-		  })
+		  }
+	)
 > {
 	static defaultProps = {
 		default: true

--- a/test/ts/refs.tsx
+++ b/test/ts/refs.tsx
@@ -6,7 +6,7 @@ import {
 	Fragment,
 	RefObject,
 	RefCallback
-} from '../../src';
+} from '../../';
 
 // Test Fixtures
 const Foo: FunctionalComponent = () => <span>Foo</span>;

